File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-rust/src/outcome.rs
// Copyright (C) 2026 Antony Stubbs and contributors

//! The user's function, and what it returns.
//!
//! **This is where Rust's lack of exceptions is answered.** Every other client library in this
//! project translates a thrown exception into a failure outcome in exactly one place; Rust has
//! nothing to translate, so the outcome is the return value: `Ok(Outcome)` is a success and
//! `Err(ProcessingError)` *is* the failure outcome. There is no `Outcome::failure()` constructor
//! to keep in step with the error path, and no way to express both at once - the type system
//! enforces what other languages enforce by convention.

use std::fmt;
use std::future::Future;
use std::sync::Arc;

use crate::proto;
use crate::record::{InboundRecord, OutboundRecord};

/// What a successful invocation produced: nothing, or records for the proxy to produce.
///
/// Failure is not a variant of this type, because `Result` already is one.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Outcome {
    produce: Vec<OutboundRecord>,
}

impl Outcome {
    /// The record was processed, with no output.
    #[must_use]
    pub fn success() -> Self {
        Self::default()
    }

    /// The record was processed, and the proxy should produce these records with its own producer
    /// before the input record's offset may become eligible to commit. This is the only sanctioned
    /// route for worker output to Kafka.
    #[must_use]
    pub fn produce(records: impl IntoIterator<Item = OutboundRecord>) -> Self {
        Self {
            produce: records.into_iter().collect(),
        }
    }

    pub(crate) fn into_wire(self) -> Vec<proto::ProduceRecord> {
        self.produce.into_iter().map(OutboundRecord::into_wire).collect()
    }
}

/// The user function's failure: the reason that rides back with the record's redelivery.
///
/// It converts from any standard error, so `?` is the whole failure path inside a processor. It
/// deliberately does **not** implement [`std::error::Error`] itself - that is what keeps the
/// blanket conversion legal, and it is the same trade [`anyhow::Error`](https://docs.rs/anyhow)
/// makes.
///
/// The reason text reaches the proxy's logs and the next delivery: **do not put record payload or
/// credentials in it.**
#[derive(Clone, PartialEq, Eq)]
pub struct ProcessingError {
    reason: String,
}

impl ProcessingError {
    /// A failure with this reason.
    #[must_use]
    pub fn new(reason: impl Into<String>) -> Self {
        Self { reason: reason.into() }
    }

    /// The reason text, as it will travel on the wire.
    #[must_use]
    pub fn reason(&self) -> &str {
        &self.reason
    }

    pub(crate) fn into_reason(self) -> String {
        self.reason
    }
}

impl fmt::Display for ProcessingError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.reason)
    }
}

impl fmt::Debug for ProcessingError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_tuple("ProcessingError").field(&self.reason).finish()
    }
}

/// The blanket conversion that makes `?` the failure path.
///
/// There is deliberately no `From<String>` or `From<&str>` beside it: the compiler refuses both,
/// because a future release of the standard library could implement [`std::error::Error`] for
/// either. [`ProcessingError::new`] is the spelling for a bare message, and it reads better at the
/// call site than a conversion would.
impl<E: std::error::Error + Send + Sync + 'static> From<E> for ProcessingError {
    fn from(error: E) -> Self {
        Self::new(error.to_string())
    }
}

/// The user's function.
///
/// A trait rather than only a closure type, because the surface must work for a language with no
/// closures at all - and because a processor that owns state (a connection pool, a counter) is the
/// ordinary case. Closures get there for free: any `Fn(InboundRecord) -> impl Future` implements
/// this, so `client.poll(|record| async move { ... })` needs no adapter.
///
/// Implementations are shared across every executor, so they take `&self` and must be `Sync`. A
/// **blocking** function belongs in [`blocking`], not here: this method is polled on the async
/// runtime, and blocking it would stall the transport task that has to keep reading the stream.
pub trait RecordProcessor: Send + Sync + 'static {
    /// Processes one record. `Err` is the failure outcome, reported with its reason.
    fn process(&self, record: InboundRecord) -> impl Future<Output = Result<Outcome, ProcessingError>> + Send;
}

impl<F, Fut> RecordProcessor for F
where
    F: Fn(InboundRecord) -> Fut + Send + Sync + 'static,
    Fut: Future<Output = Result<Outcome, ProcessingError>> + Send,
{
    fn process(&self, record: InboundRecord) -> impl Future<Output = Result<Outcome, ProcessingError>> + Send {
        self(record)
    }
}

/// A processor built from an ordinary **blocking** function - a database call, a synchronous HTTP
/// client, a CPU-bound transform.
///
/// It runs each invocation on the runtime's blocking pool, so user code that blocks cannot stall
/// the task that reads the session stream. Reaching for this is the right answer whenever the
/// function is not already `async`; wrapping blocking work in an `async` block is the Rust mistake
/// this exists to prevent.
///
/// ```no_run
/// # use parallel_consumer_proxy_client::{blocking, Outcome, ParallelConsumerClient};
/// # fn go(client: &ParallelConsumerClient) -> Result<(), Box<dyn std::error::Error>> {
/// client.poll(blocking(|record| {
///     std::fs::write("/tmp/last-record", record.value.unwrap_or_default())?;
///     Ok(Outcome::success())
/// }))?;
/// # Ok(()) }
/// ```
pub fn blocking<F>(function: F) -> Blocking<F>
where
    F: Fn(InboundRecord) -> Result<Outcome, ProcessingError> + Send + Sync + 'static,
{
    Blocking(Arc::new(function))
}

/// The processor [`blocking`] returns.
#[derive(Debug)]
pub struct Blocking<F>(Arc<F>);

impl<F> RecordProcessor for Blocking<F>
where
    F: Fn(InboundRecord) -> Result<Outcome, ProcessingError> + Send + Sync + 'static,
{
    fn process(&self, record: InboundRecord) -> impl Future<Output = Result<Outcome, ProcessingError>> + Send {
        let function = Arc::clone(&self.0);
        async move {
            match tokio::task::spawn_blocking(move || function(record)).await {
                Ok(outcome) => outcome,
                // A panic in the user's function is a failure report, never a torn-down session.
                Err(join) => Err(ProcessingError::new(crate::client::panic_reason(join))),
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Debug)]
    struct Unreachable;

    impl fmt::Display for Unreachable {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            f.write_str("the downstream service was unreachable")
        }
    }

    impl std::error::Error for Unreachable {}

    #[test]
    fn any_standard_error_becomes_a_failure_reason() {
        fn user_code() -> Result<Outcome, ProcessingError> {
            Err(Unreachable)?
        }

        assert_eq!(
            user_code().unwrap_err().reason(),
            "the downstream service was unreachable"
        );
    }

    /// The surface must work for a caller that cannot write a closure, so a named type has to be
    /// able to implement the trait alongside the blanket closure impl. If the two ever collide,
    /// this stops compiling - which is the finding, not a nuisance.
    #[test]
    fn a_named_type_can_be_a_processor() {
        struct Counter;

        impl RecordProcessor for Counter {
            async fn process(&self, _record: InboundRecord) -> Result<Outcome, ProcessingError> {
                Ok(Outcome::success())
            }
        }

        fn accepts<P: RecordProcessor>(_processor: P) {}

        accepts(Counter);
        accepts(|_record: InboundRecord| async { Ok(Outcome::success()) });
        accepts(blocking(|_record: InboundRecord| Ok(Outcome::success())));
    }
}

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-dotnet/src/Bz.Stub.ParallelConsumer.Proxy.Client/Outcome.cs
// Copyright (C) 2026 Antony Stubbs and contributors

using Google.Protobuf;
using Bz.Stub.ParallelConsumer.Proxy.Protocol.V1;

namespace Bz.Stub.ParallelConsumer.Proxy.Client;

/// <summary>
/// A record the user's function asks the proxy to produce on success.
/// </summary>
/// <remarks>
/// Workers never touch Kafka themselves. Output rides the success report and the proxy produces it
/// with its own producer, before the input record's offset may become eligible to commit.
/// </remarks>
/// <param name="Topic">The destination topic.</param>
/// <param name="Key">The key, or <see langword="null"/> for a null key.</param>
/// <param name="Value">The value, or <see langword="null"/> for a tombstone.</param>
public sealed record OutboundRecord(string Topic, byte[]? Key = null, byte[]? Value = null)
{
    internal ProduceRecord ToProduceRecord()
    {
        var produce = new ProduceRecord { Topic = Topic };
        if (Key is not null)
        {
            produce.Key = ByteString.CopyFrom(Key);
        }

        if (Value is not null)
        {
            produce.Value = ByteString.CopyFrom(Value);
        }

        return produce;
    }
}

/// <summary>
/// What the user's function decided about one record: it succeeded (optionally producing records),
/// or it failed with a reason.
/// </summary>
/// <remarks>
/// C# has exceptions, so there are two spellings of failure and exactly one meaning: returning
/// <see cref="Fail"/> and throwing are both reported as a <c>Failure</c> outcome, and the
/// translation from a thrown exception happens once, in one place
/// (<see cref="ParallelConsumerClient"/>). Nothing else in this library catches a processor's
/// exception.
/// </remarks>
public readonly record struct Outcome
{
    private Outcome(string? failureReason, IReadOnlyList<OutboundRecord> produce)
    {
        FailureReason = failureReason;
        Produce = produce;
    }

    /// <summary>Whether the record was processed successfully.</summary>
    public bool IsSuccess => FailureReason is null;

    /// <summary>
    /// The failure text when this is a failure, otherwise <see langword="null"/>. It rides back on
    /// the redelivery as <see cref="InboundRecord.LastFailureReason"/> and reaches the proxy's
    /// logs: do not put record payload or credentials in it.
    /// </summary>
    public string? FailureReason { get; }

    /// <summary>Records the proxy should produce on success. Empty means produce nothing.</summary>
    public IReadOnlyList<OutboundRecord> Produce { get; }

    /// <summary>The record was processed. Nothing to produce.</summary>
    public static Outcome Succeed() => new(null, Array.Empty<OutboundRecord>());

    /// <summary>
    /// The record was processed, and the proxy should produce these records with its own producer.
    /// This is the only sanctioned route for worker output to Kafka.
    /// </summary>
    /// <param name="records">The records to produce.</param>
    public static Outcome SucceedProducing(params OutboundRecord[] records) =>
        new(null, records ?? Array.Empty<OutboundRecord>());

    /// <summary>
    /// The record failed. It returns to the engine's retry scheduling, exactly as an in-process
    /// user function throwing would leave it.
    /// </summary>
    /// <param name="reason">Why it failed. Worker-supplied text; see <see cref="FailureReason"/>.</param>
    public static Outcome Fail(string reason) =>
        new(reason ?? string.Empty, Array.Empty<OutboundRecord>());
}

/// <summary>
/// The user's function: take a record, do the work, say what happened.
/// </summary>
/// <remarks>
/// A thrown exception is translated to a failure outcome, so an application that already signals
/// failure by throwing needs no adapter; <see cref="Outcome.Fail"/> is for the case where failure
/// is an expected result rather than an exceptional one.
/// <para>
/// The <see cref="CancellationToken"/> is the application's, not a per-record deadline: this
/// protocol never builds one, because the lease is connection liveness rather than a time budget
/// for the user's function.
/// </para>
/// </remarks>
/// <param name="record">The record to process.</param>
/// <param name="cancellationToken">The application's cancellation token.</param>
/// <returns>The outcome to report.</returns>
public delegate ValueTask<Outcome> RecordProcessor(InboundRecord record, CancellationToken cancellationToken);
